1846 support solargraph lsp

hsanson opened the pull request:
Adds support for solargraph language server.

Closes #1846

w0rp commented on ale_linters/ruby/solargraph.vim:
```diff
@@ -0,0 +1,19 @@
+
+function! ale_linters#ruby#solargraph#GetAddress(buffer) abort
+    let l:host = ale#VarFunc('ruby_solargraph_host')(a:buffer)
+    let l:port = ale#VarFunc('ruby_solargraph_port')(a:buffer)
```
Use `ale#Var(a:buffer, variable_name)` instead. `VarFunc` is just for writing simple callbacks easily.

w0rp commented on autoload/ale/ruby.vim:
```diff
@@ -20,3 +20,65 @@ function! ale#ruby#FindRailsRoot(buffer) abort
+    let l:dir = fnamemodify(
+    \   ale#path#FindNearestFile(a:buffer, "Rakefile"),
+    \   ':h'
+    \)
```
You're searching for the same two files twice here.

w0rp commented on autoload/ale/ruby.vim:
```diff
@@ -20,3 +20,65 @@ function! ale#ruby#FindRailsRoot(buffer) abort
+
+    if isdirectory(l:dir)
+      return l:dir
+    endif
```
Add `return ''` to the end, or the function implicitly returns `0`.

hsanson commented:
@w0rp I have added tests, simplified the find root logic, and updated the documentation as requested.

BTW: the dev documentation does not mention anything about rebase policies. I rebased master and squashed everything into a single commit for easier review. Hope this is not a problem as I had to force push for this to work.

w0rp commented:
There is no policy. I just merge code that's good and use `--squash` if the commits are messy.

w0rp commented:
Cheers! :beers:

w0rp commented:
I might add an option to disable TCP linters by default later if it turns out that trying and failing to make connections causes an issue. We'll see what happens.
